### DISCUSSION TITLE
Is it possible for state names to be type safe? (v5)

### DISCUSSION THREAD

**Discussion by realfresh** [statelyai/xstate#4382] [Q&A]
I'm using xstate v5, havent used any prior versions. It's extremely frustrating that the names of a state are not typesafe.

Consider this code:
```
const toggleMachine = createMachine({
  id: "toggle",
  context: {count: 0},
  initial: "DoesntExist",
  states: {
    Inactive: {
      on: {toggle: "DoesntExist"},
    },
    Active: {
      entry: assign({count: ({context}) => context.count + 1}),
      on: {toggle: "DoesntExist"},
      after: {2000: "DoesntExist"},
    },
  },
});
```

I can start at a state that doesn't exist and transition to states that don't exist.

I've tried every possibly way to ensure strongly typed state names like:
```
types: {
  state: "Inactive" | "Active";
}

// OR

types: {
  state: {
    Inactive: {},
    Active: {},
  }
}
```

None of that works. I even resorted to doing:
```
const states = { Inactive: "Inactive", Active: "Active" } as const;

const toggleMachine = createMachine({
  states: {
    [states.Inactive]: {...},
    [states.Active]: {...},
  },
});
```

This is good, but then the carts at stately.ai don't work anymore with that pattern.

I mean, is this by design? And is there anyway around this? 

It's too easy to make mistakes with state names in a large state machine when refactoring

**Comment by Andarist** [comment]
IIRC, our VS Code extension can currently raise warnings about invalid state names. At the moment, we can't provide more than that **but** we are actively looking for improvements in this area. The upcoming TS 5.3 (scheduled to be released in November) might also allow us to do some new neat things and we'll definitely be looking for improvements that it might unlock here.

**Reply by stanlrt** [reply]
Any news?

**Reply by davidkpiano** [reply]
Yes! State names are type-safe if you use `setup({…}).createMachine({…})`
